### PR TITLE
50085449 integrating list with map

### PR DESCRIPTION
**Pull request by OlenaCh**
Fixes: https://www.pivotaltracker.com/story/show/50085449. 
Integrating list of existing volops with the Google map (also, adding relevant RSpec & Cucumber tests).
P. S. After first checks review, it looks like this PR will require a little more additional work.

### CODE REVIEW

**Comment by tansaku** [comment]
thanks for getting started on this @OlenaCh - might it be possible to split the work up into stages that can be merged separately? or make the work itself simpler?

**Comment by OlenaCh** [comment]
@tansaku you're welcome! Actually, I'm not sure if it is possible - the main changes were applied to maps.js and other ones simply help to provide js starting function with necessary data. However, I've made a mistake to organize methods in VolunteerOp model by alphabet (it was easier to look for a method) and due to this committed changes look so enormous. I think I'll return to previous structure to make it easier to navigate.

**Comment by OlenaCh** [comment]
@tansaku may I ask you about test coverage, please? I was fixing failed checks and found out a couple of issues.

First, there is a button "This is my organisation" which contains link to organiation and, also, is hooked to a JS event onclick. As a result, Cucumber fails when it gets to point where it is necessary to test how sign in form can be opened with a script. Instead of opening that form Cucumber gets link from button and follows it. Is it possible somehow to fix it (I've google the issue, but found no effective solution)? 

Second, there is an issue with RSpec tests. When I've started working on adding more coverage, I looked through registration controller and users controller code (which were lacking it) and found out that some methods duplicate each other. So:
 
**UsersController**
def _update_message_for_admin_status_
    org = Organisation.friendly.find(params[:pending_organisation_id])
    flash[:notice] = "You have requested admin status for #{org.name}"
    send_email_to_superadmin_about_request_for_admin_of org   
    redirect_to organisation_path(org)
end

def _send_email_to_superadmin_about_request_for_admin_of org_
    superadmin_emails = User.superadmins.pluck(:email)
    AdminMailer.new_user_waiting_for_approval(org.name, superadmin_emails).deliver_now
end
  
**RegistrationController**
def _update_message_for_admin_status_
    org = Organisation.find(params[:user][:pending_organisation_id])
    flash[:notice] << " You have requested admin status for #{org.name}"
    send_email_to_superadmin_about_request_for_admin_of org
end

def _send_email_to_superadmin_about_request_for_admin_of org_
  superadmin_emails = User.superadmins.pluck(:email)
  AdminMailer.new_user_waiting_for_approval(org.name, superadmin_emails).deliver_now
end

May I ask you what I should do in this case? I was not working on those controllers logic, so I'm not sure if it is wise to refactor them. Still, I can try to cover more of registration controller with tests. 

Thank you very much!

File: app/assets/javascripts/application.js [hunk 1]
```diff
@@ -10,6 +10,7 @@
 //= require underscore
 //= require bootstrap-sortable
 //= require_tree ./global
+//= require_tree
```
Comment (Maroo-b) [inline app/assets/javascripts/application.js]: why did you add `require_tree` ?

File: app/models/volunteer_op.rb [hunk 2]
```diff
@@ -14,7 +14,6 @@ class VolunteerOp < ActiveRecord::Base
   scope :reachskills,          -> { where(source: 'reachskills') }
   scope :remote_only,          -> { where.not(source: 'local') }
```
Comment (Maroo-b) [inline app/models/volunteer_op.rb]: would it better if you revert the sort of method by name ? To have less files to review for the moment :)

File: features/home_page/maintain_org_pos_in_list.feature [hunk 3]
```diff
@@ -4,11 +4,10 @@ Feature: maintain position in organisation list
...
   Background: organisations have been added to database
-    
     Given the following organisations exist:
       | name                           | description                    | address        | postcode | website       |
       | Harrow Bereavement Counselling | Harrow Bereavement Counselling | 34 Pinner Road | HA1 4HZ  | http://a.com/ |
 
   Scenario:
     When I visit the organisations index page
-    Then I should see link "Harrow Bereavement Counselling" targeting new page
+    Then I should see link "Harrow Bereavement Counselling" targeting new page
```
Comment (Maroo-b) [inline features/home_page/maintain_org_pos_in_list.feature]: I noticed that you sometimes add return line to end of statements. Maybe you could configure your editor to avoid this case :)

**Comment by Maroo-b** [comment]
@OlenaCh regarding the failing Cucumber test for "This is my organisation" case, is it an existing test ? Generally when testing Javascript interaction via Cucumber we need a javascript driver in this project we are using webkit: check this config ([link](https://github.com/AgileVentures/LocalSupport/blob/develop/features/support/env.rb#L37))

> May I ask you what I should do in this case? I was not working on those controllers logic, so I'm not sure if it is wise to refactor them. Still, I can try to cover more of registration controller with tests.

for your next question I suggest that you open a refactoring ticket on Pivotal Tracker (PT), and you can focus only on the current ticket :)

File: app/assets/javascripts/application.js (hunk 1 above)
Comment (OlenaCh) [inline app/assets/javascripts/application.js]: Oh, sorry, it's force of habit - I've removed it already :) .

File: features/home_page/maintain_org_pos_in_list.feature (hunk 3 above)
Comment (OlenaCh) [inline features/home_page/maintain_org_pos_in_list.feature]: Oh, of course! Thank you :)

File: app/models/volunteer_op.rb (hunk 2 above)
Comment (OlenaCh) [inline app/models/volunteer_op.rb]: Should I change this file to its previous state without sorting methods by name? Or should I keep the sorting? Sorry for asking; before you commented I'd already returned to previous state without sorting and now just switched back. Anyway, I'll redo it as necessary.

**Comment by OlenaCh** [comment]
@Maroo-b sorry for a little delay (I got a cold and was not in a good working condition :) ). 
Looks like, the Cucumber tests problem was, indeed, about configuration. Thank you very much for help! 
Additionally, I fixed Cucumber test about reachskills org opening in a new window. Currently, when user clicks on org's title in the volunteer ops list, a relevant info box should be opened on the Google map. So, I've commented that scenario about opening in a new window and added a new one about opening info box. If later some new changes are introduced, it would be easy to switch between these tests.
Also, I'll create a new Pivotal Tracker issue about RSpec, too. 
Thank you very much for comments and help!

File: app/assets/stylesheets/gmaps.css.scss [hunk 4]
```diff
@@ -124,6 +124,21 @@
   }
 }
 
+strong.organisation-title {
+  color: #006937;
```
Comment (Maroo-b) [inline app/assets/stylesheets/gmaps.css.scss]: for color values I suggest to use the following color [file](https://github.com/AgileVentures/LocalSupport/blob/develop/app/assets/stylesheets/utils/_colors.css.scss), it will be helpful in future to change color degrees or values from one location.

File: app/models/volunteer_op.rb (hunk 2 above)
Comment (Maroo-b) [inline app/models/volunteer_op.rb]: Well personally I prefer to see real code changes in git diff instead of sorting, I agree that sorting could be helpful but in case of new feature implementation let's keep unecessary changes to the minimum to have more focus on the implementation :)

File: spec/controllers/users_controller_spec.rb [hunk 5]
```diff
@@ -1,14 +1,46 @@
...
+                                      pending_organisation_id: 5
+                                    }.with_indifferent_access)
     end
   end
-end
+  
+  describe 'edit' do
+    before(:each) { get :edit, id: user.id }
+    
+    it 'assigns @volunteer_ops variable' do
```
Comment (Maroo-b) [inline spec/controllers/users_controller_spec.rb]: I guess you meant **'assigns @user variable'** as description.

File: spec/controllers/users_controller_spec.rb [hunk 6]
```diff
@@ -1,14 +1,46 @@
 require 'rails_helper'
 
 describe UsersController, :type => :controller do
-  describe ".permit" do
-    it "returns the cleaned params" do
-      user_params = {id: 5, email: "a@a.com", password: "blahblah", password_confirmation: "blahblah", 
-       remember_me: true, pending_organisation_id: 5, superadmin: true}
+  let(:user) { FactoryGirl.create(:user) }
```
Comment (Maroo-b) [inline spec/controllers/users_controller_spec.rb]: just to confirm what's the goal of adding a User factory in this test ?

**Comment by Maroo-b** [comment]
@OlenaCh I hope you are better now :) 

I saw that you have made some changes to users_controller_spec file and new controller test for volunteer_ops. Usually in this project we avoid adding controller tests unless there is some cases not covered by features test.

@tansaku I noticed there is a huge number of **feature/req_cache** files is it ok or do we need additional configuration ?

**Comment by tansaku** [comment]
@Maroo-b there may be more than are needed.  The problem is that new ones can be generated due to failed test runs, but then we don't necessarily need them all ...

It's an ongoing cause of concern.  I'm still not sure what the best way to handle it is.

If I was managing it myself I would pull all the pure code changes into one branch and then do a single complete test run there and hopefully see a smaller set of changes.

That said, there can be a lot of new files associate with a new feature, particularly if we are doing it involves using maps etc. in a slightly different way ...

File: app/assets/stylesheets/gmaps.css.scss (hunk 4 above)
Comment (OlenaCh) [inline app/assets/stylesheets/gmaps.css.scss]: Ok, I'll change it.

File: app/models/volunteer_op.rb (hunk 2 above)
Comment (OlenaCh) [inline app/models/volunteer_op.rb]: Sorry, I'll avoid it in the future :) . I'll redo code as it was, but in case if it keeps showing a lot of changed lines I can also name methods which were changed. Actually, this model was not changed very much. I simply added one more method and refactored another to avoid duplication.

File: spec/controllers/users_controller_spec.rb (hunk 6 above)
Comment (OlenaCh) [inline spec/controllers/users_controller_spec.rb]: To have a user id which can be sent as a parameter later during testing edit action. Should I remove it to the relevant test case?

File: spec/controllers/users_controller_spec.rb (hunk 5 above)
Comment (OlenaCh) [inline spec/controllers/users_controller_spec.rb]: Oh, yes, I'll correct it :) . Thank you!

**Comment by OlenaCh** [comment]
@Maroo-b yes, I'm, thank you :). I knew about that rule and I added those tests only because there would be a drop in test coverage later during PR checks. At the beginning I had added volunteer ops controller tests, but even in that case I was not able to pass my first PR checks without adding users controller tests. As for now, I can remove those tests and leave only relevant Cucumber tests to be.
@tansaku maybe, I can try and do this, too. However, I do not understand very well what should be done. Should I create one more branch which would contain only JS and Ruby code changes without tests?

**Comment by Maroo-b** [comment]
@OlenaCh did you finish the implementation to check this feature locally ? :)

**Comment by OlenaCh** [comment]
@Maroo-b yes, it can be tested locally. I'll review code one more time in the evening (because of newly added scope - reachskills), but it should work well.

**Comment by OlenaCh** [comment]
**UPDATED!** @Maroo-b yes, it can be tested locally. I've reviewed code one more time and it had been working well. 

Also, I've removed those RSpec tests which I added previously (consequently, there are no volunteer ops controller spec and a couple of tests in the users controller spec). I've to admit I do not understand why this time there was no issue with failing test coverage check, but anyway that's great :) ! 

Also, may I ask you, please, what I should do further with branches? I was thinking that I could create one more branch, add only code changes, applied to JS and Ruby code (also, it would help with volunteer op model method sorting issue), and redo a PR. After that, one more PR would be only about Cucumber tests. Or is it unnecessary?

**Comment by tansaku** [comment]
good chatting with you about this feature in https://www.youtube.com/watch?v=PFkVKvQ8Ze4 @OlenaCh
